include/sound/sof/ext_manifest4.h

```diff
@@ -0,0 +1,135 @@
+/* SPDX-License-Identifier: ((GPL-2.0 WITH Linux-syscall-note) OR BSD-3-Clause) */
+/*
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2021 Intel Corporation. All rights reserved.
+ */
+
+/*
+ * Extended manifest is a place to store metadata about firmware, known during
+ * compilation time - for example firmware version or used compiler.
+ * Given information are read on host side before firmware startup.
+ * This part of output binary is not signed.
+ */
+
+#ifndef __SOF_FIRMWARE_EXT_MANIFEST4_H__
+#define __SOF_FIRMWARE_EXT_MANIFEST4_H__
+
+/* In ASCII  $AE1 */
+#define SOF_EXT_MAN4_MAGIC_NUMBER    0x31454124
+
+#define MAX_MODULE_NAME_LEN     8
+#define MAX_FW_BINARY_NAME      8
+#define DEFAULT_HASH_SHA256_LEN 32
+#define SOF_MAN4_FW_HDR_OFFSET	0x2000
+
+/*********************************************************************
+ *     css_manifest hdr
+ *-------------------
+ *     offset reserved for future
+ *-------------------
+ *     fw_hdr
+ *-------------------
+ *     module_entry[0]
+ *-------------------
+ *     module_entry[1]
+ *-------------------
+ *     ...
+ *-------------------
+ *     module_entry[n]
+ *-------------------
+ *     FW content
+ *-------------------
+ *********************************************************************/
+
+struct sof_ext_manifest4_hdr {
+	u32 id;
+	u32 len;
+	u16 version_major;
+	u16 version_minor;
+	u32 entries;
+} __packed;
+
+struct sof_man4_fw_img_flags {
+	uint32_t _rsvd0 : 31;
+	uint32_t tp : 1;
+} __packed;
+
+struct sof_man4_fw_binary_header {
+	/* This part must be unchanged to be backward compatible with SPT-LP ROM */
+	uint32_t		id;
+	uint32_t		len; /* sizeof(sof_man4_fw_binary_header) in bytes */
+	uint8_t			name[MAX_FW_BINARY_NAME];
+	uint32_t		preload_page_count; /* number of pages of preloaded image */
+	struct sof_man4_fw_img_flags	fw_image_flags;
+	uint32_t		feature_mask;
+	uint16_t		major_version;
+	uint16_t		minor_version;
+	uint16_t		hotfix_version;
+	uint16_t		build_version;
+	uint32_t		num_module_entries;
+
+	/* This part may change to contain any additional data for BaseFw that is skipped by ROM */
+	uint32_t		hw_buf_base_addr;
+	uint32_t		hw_buf_length;
+	uint32_t		load_offset; /* This value is used by ROM */
+} __packed;
+
+struct sof_man4_module_type {
+	uint32_t load_type : 4; /* MT_BUILTIN, MT_LOADABLE */
+	uint32_t auto_start : 1; /* 0 - manually created, 1 - created by Module Manager */
+	uint32_t domain_ll : 1; /* support LL domain */
+	uint32_t domain_dp : 1; /* support DP domain */
+	uint32_t lib_code : 1; /* determines if module is place holder for common library code */
+	uint32_t _rsvd : 24;
+} __packed;
```

Comment:
It might be only me, but parsing the bit definition like this is not too user and compiler friendly.
With gcc the `struct sof_man4_module_type` is four bytes (uint32), and not sure what the __packed supposed to do.
For a quick test:
```c
#include <stdio.h>

typedef unsigned int uint32_t;

struct sof_man4_module_type_1 {
	uint32_t load_type; /* MT_BUILTIN, MT_LOADABLE */
	uint32_t auto_start; /* 0 - manually created, 1 - created by Module Manager */
	uint32_t domain_ll; /* support LL domain */
	uint32_t domain_dp; /* support DP domain */
	uint32_t lib_code; /* determines if module is place holder for common library code */
	uint32_t _rsvd;
};

struct sof_man4_module_type_2 {
	uint32_t load_type : 4; /* MT_BUILTIN, MT_LOADABLE */
	uint32_t auto_start : 1; /* 0 - manually created, 1 - created by Module Manager */
	uint32_t domain_ll : 1; /* support LL domain */
	uint32_t domain_dp : 1; /* support DP domain */
	uint32_t lib_code : 1; /* determines if module is place holder for common library code */
	uint32_t _rsvd : 24;
};

struct sof_man4_module_type_3 {
	uint32_t load_type : 4; /* MT_BUILTIN, MT_LOADABLE */
	uint32_t auto_start : 1; /* 0 - manually created, 1 - created by Module Manager */
	uint32_t domain_ll : 1; /* support LL domain */
	uint32_t domain_dp : 1; /* support DP domain */
	uint32_t lib_code : 1; /* determines if module is place holder for common library code */
	uint32_t _rsvd : 24;
} __packed;

int main(int argc, char *argv[])
{
	struct sof_man4_module_type_1 type_1;
	struct sof_man4_module_type_2 type_2;
	struct sof_man4_module_type_3 type_3;

	printf( "sizeof sof_man4_module_type_1 : %lu\n", sizeof(type_1));
	printf( "sizeof sof_man4_module_type_2 : %lu\n", sizeof(type_2));
	printf( "sizeof sof_man4_module_type_3 : %lu\n", sizeof(type_3));

	return 0;
}
```
Results:
```
sizeof sof_man4_module_type_1 : 24
sizeof sof_man4_module_type_2 : 4
sizeof sof_man4_module_type_3 : 4
```

Reply:
@ujfalusi The bit definition allocation is from firmware side. We can only use either `struct sof_man4_module_type` or `uint32_t sof_man4_module_type`. And it is easier to access the `load_type` field by `type.load_type` than using a macro to me.
Regarding `__packed`, it is to avoid the structures to be padded. There is no different between your `struct sof_man4_module_type_2` and `struct sof_man4_module_type_3` is because there is only one item in the structure. You can see the size of sof_man4_module_type_2 is 8 and sof_man4_module_type_3 is 5 if you add a `char` item in the structure.
BTW, I have to use ` __attribute__((__packed__))` instead of `__packed` when I tested it. But `__packed` is preferred in kernel space.

Reply:
@bardliao, right, but consider if the compiler for a given platform is handling the C bitfield in a different way (or not at all).
Also if you see a hex dump of `0x00000053`, how would you quickly check what it is?
```c
struct sof_man4_module_type {
	uint32_t load_type : 4; /* MT_BUILTIN, MT_LOADABLE */
	uint32_t auto_start : 1; /* 0 - manually created, 1 - created by Module Manager */
	uint32_t domain_ll : 1; /* support LL domain */
	uint32_t domain_dp : 1; /* support DP domain */
	uint32_t lib_code : 1; /* determines if module is place holder for common library code */
	uint32_t _rsvd : 24;
} __packed;
```
VS
```c
#define SOF_MAN4_MODULE_LOAD_TYPE	GENMASK(0, 3)
#define SOF_MAN4_MODULE_AUTO_START 	BIT(4)
#define SOF_MAN4_MODULE_DOMAIN_LL	BIT(5)
#define SOF_MAN4_MODULE_DOMAIN_DP	BIT(6)
#define SOF_MAN4_MODULE_LIB_CODE	BIT(7)
```
If we also want to debug this at some point then probably do something like:
```c
union sof_man4_module_type {
	uint32_t as_uint32;
	struct {
		uint32_t load_type : 4; /* MT_BUILTIN, MT_LOADABLE */
		uint32_t auto_start : 1; /* 0 - manually created, 1 - created by Module Manager */
		uint32_t domain_ll : 1; /* support LL domain */
		uint32_t domain_dp : 1; /* support DP domain */
		uint32_t lib_code : 1; /* determines if module is place holder for common library code */
		uint32_t _rsvd : 24;
	} __packed;
};
```

Reply:
@ujfalusi You convinced me, I will use macros. Thanks.